erizo/src/erizo/rtp/BandwidthEstimationHandler.cpp

```diff
@@ -98,24 +98,203 @@ void BandwidthEstimationHandler::updateExtensionMap(bool is_video, std::array<RT
+    // RTP PACKET
     if (parsePacket(packet)) {
+      webrtc::StatisticianMap res = _stats_handler->GetActiveStatisticians();   // get a list of active stats engine
+      webrtc::StatisticianMap::iterator it;
+
       int64_t arrival_time_ms = packet->received_time_ms;
       arrival_time_ms = clock_->TimeInMilliseconds() - (ClockUtils::timePointToMs(clock::now()) - arrival_time_ms);
       size_t payload_size = packet->length;
-      pickEstimatorFromHeader();
-      rbe_->IncomingPacket(arrival_time_ms, payload_size, header_);
+      if (packet->type == VIDEO_PACKET) {
+        // SEND TO BWE
+        pickEstimatorFromHeader();
+        rbe_->IncomingPacket(arrival_time_ms, payload_size, header_);
+
+        // SEND TO OUR STATS ENGINE
+        it = res.find(videoRR.ssrc);  // pick our video stats engine
+        bool isRetransmitted = false;
+        if (it != res.end()) {
+          // CONSIDER TO ADD MORE PAYLOAD TYPE
+          header_.payload_type_frequency = 90000;
+          isRetransmitted = it->second->IsRetransmitOfOldPacket(header_, 0);   // check if packed is a retransmitted one
+        }                                                                      // to make our stats more accurate
+        _stats_handler->IncomingPacket(header_, packet->length, isRetransmitted);
+
+      } else if (packet->type == AUDIO_PACKET) {
+        // SEND TO OUR STATS ENGINE
+        it = res.find(audioRR.ssrc);  // pick our audio stats engine
+        bool isRetransmitted = false;
+        int clockRate = 48000;
+        if (it != res.end()) {
+          // CONSIDER TO ADD MORE PAYLOAD TYPE
+          if (head->getPayloadType() == OPUS_48000_PT) {
+            clockRate = 48000;
+          } else {
+            clockRate = 8000;
+          }
+          header_.payload_type_frequency = clockRate;
+          isRetransmitted = it->second->IsRetransmitOfOldPacket(header_, 0);  // check if packed is a retransmitted one
+        }                                                                     // to make our stats more accurate
+        _stats_handler->IncomingPacket(header_, packet->length, isRetransmitted);
+      }
+
     } else {
       ELOG_DEBUG("Packet not parsed %d", packet->type);
     }
   }
+
+  // RR PACKET HANDLER - generate RR packets as per rfc3550
+  // TODO(kekkokk) Calculate the right timing when send RR based on available bandwidth
+  // For semplicity an RR packet is sent every SR packet received.
+
+  if (!chead->isRtcp()) {
+    // RTP PACKETS
+    uint16_t seq_num = head->getSeqNumber();
+
+    switch (packet->type) {
+      case VIDEO_PACKET: {
+        // RTP VIDEO PACKET
+        // CALCULATE CYCLE
+        videoRR.ssrc = head->getSSRC();
+        if (!rtpSequenceLessThan(seq_num, videoRR.max_seq)) {
+          if (seq_num < videoRR.max_seq) {
+            videoRR.cycle++;
+          }
+          videoRR.max_seq = seq_num;
+        }
+        break;
+      }
+      case AUDIO_PACKET: {
+        // RTP AUDIO PACKET
```

Comment:
please, remove this kind of comments since it's kind of obvious to handle a RTP AUDIO PACKET inside an AUDIO_PACKET case. I don't think CALCULATE CYCLE is needed either or you can use a generic function called calculateCycle(...)